Remove 'minute' frequency option from SLM policy form because ES won't allow a frequency faster than every 15 minutes.

cjcenizal opened the pull request:
## Summary

Fixes https://github.com/elastic/kibana/issues/83918. If you try to create a policy using this option, ES will just reject it so it doesn't make sense to offer it to the user.

Other changes:
- Added `frequencyBlockList` prop to `CronEditor` to allow selectively removing frequency options.
- Added unit test coverage for `CronEditor` component.

![image](https://user-images.githubusercontent.com/1238659/100949485-3c4cdc00-34bf-11eb-8993-02643f3a26d0.png)

### Checklist

Delete any items that are not applicable to this PR.

- [ ] ~Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)~
- [ ] ~[Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials~
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] ~Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))~
- [ ] ~Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))~
- [ ] ~This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))~
- [ ] ~This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)~

### For maintainers

- [ ] ~This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)~

elasticmachine commented:
Pinging @elastic/es-ui (Team:Elasticsearch UI)

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91597/)
* Commit: 5a48629097963bb24e823791c244b70ba1ab95be
* This comment will update when the build is complete

### Failed CI Steps
* [Check Types](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91597/execution/node/377/log)

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91597/","number":"91597","commit":"5a48629097963bb24e823791c244b70ba1ab95be"}]}
PIPELINE-->

alisonelizabeth commented on x-pack/plugins/snapshot_restore/public/application/components/policy_form/steps/step_logistics.tsx:
```diff
@@ -480,6 +480,7 @@ export const PolicyStepLogistics: React.FunctionComponent<StepProps> = ({
       ) : (
         <Fragment>
           <CronEditor
+            unitBlockList={[MINUTE]} // ES disallows a frequency faster than 15m
```
I think this is supposed to be `frequencyBlockList`

alisonelizabeth commented on src/plugins/es_ui_shared/public/components/cron_editor/cron_editor.test.tsx:
```diff
@@ -0,0 +1,138 @@
+  };
+});
+
+describe('CronEditor', () => {
```
Thanks for adding these tests!

alisonelizabeth reviewed (commented):
Thanks for working on this @cjcenizal! 

It looks like the prop passed to the `CronEditor` is incorrect, but I was able to test successfully when I fixed this (see comment in code). It also looks like there are a few linting errors in the tests; some related to the cron editor code not yet being converted to TS.

Also, after looking at the ES issue more closely, I see the change is a restriction on a frequency <15 minutes. Do you think it would be better if we left the `minute` option, but instead add client validation on the cron expression to verify the value is >15 minutes?

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91912/)
* Commit: 376d59aea7db05e97bfb07a30f344acb842cfa41
* This comment will update when the build is complete

### Failed CI Steps
* [Check Types](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91912/execution/node/371/log)

### History
* :broken_heart: [Build #91597](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91597/) failed 5a48629097963bb24e823791c244b70ba1ab95be

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91597/","number":"91597","commit":"5a48629097963bb24e823791c244b70ba1ab95be"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91912/","number":"91912","commit":"376d59aea7db05e97bfb07a30f344acb842cfa41"}]}
PIPELINE-->

cjcenizal commented:
@elasticmachine merge upstream

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/92573/)
* Commit: 17fd6ff1ac6bf5ead93de3489674704a03fb4ef2
* This comment will update when the build is complete

### Failed CI Steps
* [Check Types](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/92573/execution/node/371/log)

### History
* :broken_heart: [Build #91912](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91912/) failed 376d59aea7db05e97bfb07a30f344acb842cfa41
* :broken_heart: [Build #91597](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91597/) failed 5a48629097963bb24e823791c244b70ba1ab95be

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91597/","number":"91597","commit":"5a48629097963bb24e823791c244b70ba1ab95be"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91912/","number":"91912","commit":"376d59aea7db05e97bfb07a30f344acb842cfa41"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/92573/","number":"92573","commit":"17fd6ff1ac6bf5ead93de3489674704a03fb4ef2"}]}
PIPELINE-->

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/92575/)
* Commit: ee8f8afc1a03a87ffdf2fa761757d396327befab
* This comment will update when the build is complete

### Failed CI Steps
* [Check Types](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/92575/execution/node/365/log)

### History
* :broken_heart: [Build #92573](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/92573/) failed 17fd6ff1ac6bf5ead93de3489674704a03fb4ef2
* :broken_heart: [Build #91912](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91912/) failed 376d59aea7db05e97bfb07a30f344acb842cfa41
* :broken_heart: [Build #91597](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91597/) failed 5a48629097963bb24e823791c244b70ba1ab95be

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91597/","number":"91597","commit":"5a48629097963bb24e823791c244b70ba1ab95be"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91912/","number":"91912","commit":"376d59aea7db05e97bfb07a30f344acb842cfa41"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/92573/","number":"92573","commit":"17fd6ff1ac6bf5ead93de3489674704a03fb4ef2"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/92575/","number":"92575","commit":"ee8f8afc1a03a87ffdf2fa761757d396327befab"}]}
PIPELINE-->

cjcenizal commented on src/plugins/es_ui_shared/public/components/cron_editor/constants.ts:
```diff
@@ -0,0 +1,166 @@
+  [key in Field]?: boolean;
+};
+
+function makeSequence(min: number, max: number): number[] {
```
The bulk of this file was extracted from `cron_editor.tsx`.

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/92584/)
* Commit: 7f27e26fd2cbda9d40ee7b4c6a01238b0a3de12b
* This comment will update when the build is complete

### Failed CI Steps
* [Check Types](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/92584/execution/node/365/log)

### History
* :broken_heart: [Build #92575](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/92575/) failed ee8f8afc1a03a87ffdf2fa761757d396327befab
* :broken_heart: [Build #92573](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/92573/) failed 17fd6ff1ac6bf5ead93de3489674704a03fb4ef2
* :broken_heart: [Build #91912](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91912/) failed 376d59aea7db05e97bfb07a30f344acb842cfa41
* :broken_heart: [Build #91597](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91597/) failed 5a48629097963bb24e823791c244b70ba1ab95be

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91597/","number":"91597","commit":"5a48629097963bb24e823791c244b70ba1ab95be"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91912/","number":"91912","commit":"376d59aea7db05e97bfb07a30f344acb842cfa41"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/92573/","number":"92573","commit":"17fd6ff1ac6bf5ead93de3489674704a03fb4ef2"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/92575/","number":"92575","commit":"ee8f8afc1a03a87ffdf2fa761757d396327befab"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/92584/","number":"92584","commit":"7f27e26fd2cbda9d40ee7b4c6a01238b0a3de12b"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/92896/)
* Commit: 6e80214ed0d8a0e796fe6b89dc9dc73fa3df815f

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/1bc42b9171387bafafac757b3ee3e012c0b78d22) | [after](https://github.com/elastic/kibana/commit/6e80214ed0d8a0e796fe6b89dc9dc73fa3df815f) | diff |
| --- | --- | --- | --- |
| `esUiShared` | 208 | 210 | +2 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/1bc42b9171387bafafac757b3ee3e012c0b78d22) | [after](https://github.com/elastic/kibana/commit/6e80214ed0d8a0e796fe6b89dc9dc73fa3df815f) | diff |
| --- | --- | --- | --- |
| `snapshotRestore` | 583.8KB | 583.8KB | +30.0B |

#### Distributable file count
| id | [before](https://github.com/elastic/kibana/commit/1bc42b9171387bafafac757b3ee3e012c0b78d22) | [after](https://github.com/elastic/kibana/commit/6e80214ed0d8a0e796fe6b89dc9dc73fa3df815f) | diff |
| --- | --- | --- | --- |
| `default` | 46934 | 47694 | +760 |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/1bc42b9171387bafafac757b3ee3e012c0b78d22) | [after](https://github.com/elastic/kibana/commit/6e80214ed0d8a0e796fe6b89dc9dc73fa3df815f) | diff |
| --- | --- | --- | --- |
| `esUiShared` | 278.2KB | 278.2KB | +5.0B |
| `snapshotRestore` | 67.5KB | 67.2KB | -297.0B |
| total |  |  | -292.0B |

### History
* :broken_heart: [Build #92584](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/92584/) failed 7f27e26fd2cbda9d40ee7b4c6a01238b0a3de12b
* :broken_heart: [Build #92575](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/92575/) failed ee8f8afc1a03a87ffdf2fa761757d396327befab
* :broken_heart: [Build #92573](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/92573/) failed 17fd6ff1ac6bf5ead93de3489674704a03fb4ef2
* :broken_heart: [Build #91912](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91912/) failed 376d59aea7db05e97bfb07a30f344acb842cfa41
* :broken_heart: [Build #91597](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91597/) failed 5a48629097963bb24e823791c244b70ba1ab95be

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91597/","number":"91597","commit":"5a48629097963bb24e823791c244b70ba1ab95be"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91912/","number":"91912","commit":"376d59aea7db05e97bfb07a30f344acb842cfa41"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/92573/","number":"92573","commit":"17fd6ff1ac6bf5ead93de3489674704a03fb4ef2"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/92575/","number":"92575","commit":"ee8f8afc1a03a87ffdf2fa761757d396327befab"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/92584/","number":"92584","commit":"7f27e26fd2cbda9d40ee7b4c6a01238b0a3de12b"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/92896/","number":"92896","commit":"6e80214ed0d8a0e796fe6b89dc9dc73fa3df815f"}]}
PIPELINE-->

cjcenizal commented on src/plugins/es_ui_shared/public/components/cron_editor/cron_monthly.tsx:
```diff
@@ -18,13 +18,21 @@
+  hourOptions: EuiSelectOption[];
+  date?: string;
+  dateOptions: EuiSelectOption[];
+  onChange: ({ minute, hour, date }: { minute?: string; hour?: string; date?: string }) => void;
```
I considered reusing `FieldToValueMap` here, but decided against it. `FieldToValueMap` creates some helpful coupling between these components and `<CronEditor />` but the greater downside is that it obscures which  specific fields are actually being provided to the `onChange` callback.

cjcenizal commented:
@alisonelizabeth I converted the code to TS and left some comments to make review easier. The TS conversion is captured within a subset of the commit history so it should be possible to review it in isolation. Could you please re-review?

> Also, after looking at the ES issue more closely, I see the change is a restriction on a frequency <15 minutes. Do you think it would be better if we left the minute option, but instead add client validation on the cron expression to verify the value is >15 minutes?

I like this idea! I think it's separate from the change we need to make here though. The UI only allows you to specify a frequency of snapshotting every 1 minute. You can't change that rate, so the option isn't useful and needs to be removed.

![image](https://user-images.githubusercontent.com/1238659/101553812-4745b780-396a-11eb-8a61-f0e25372a2fa.png)

I think it's technically possible to validate that the "minute" part of the underlying cron is greater than 15, but I don't want to block this improvement on that work so I extracted it into https://github.com/elastic/kibana/issues/85359.

alisonelizabeth reviewed (approved):
LGTM. Tested locally again and everything works as expected. Nice job converting the cron editor to TS!

> I like this idea! I think it's separate from the change we need to make here though. The UI only allows you to specify a frequency of snapshotting every 1 minute. You can't change that rate, so the option isn't useful and needs to be removed.

👍 When I originally made the comment I was thinking you could specify a number when you selected the `minute` frequency.
